napari/layers/labels/labels.py

```diff
@@ -682,46 +734,64 @@ def paint(self, coord, new_label, refresh=True):
-                        ).astype(int),
-                        1,
-                    )
-                    for c, s in zip(coord, self.shape)
-                ]
-            )
-        else:
-            slice_coord = [0] * self.ndim
-            for i in self.dims.displayed:
-                slice_coord[i] = slice(
-                    np.round(
-                        np.clip(
-                            coord[i] - self.brush_size / 2 + 0.5,
-                            0,
-                            self.shape[i],
-                        )
-                    ).astype(int),
-                    np.round(
-                        np.clip(
-                            coord[i] + self.brush_size / 2 + 0.5,
-                            0,
-                            self.shape[i],
-                        )
-                    ).astype(int),
+                slice(
+                    np.round(np.clip(c - brush_size / 2 + 0.5, 0, s)).astype(
+                        int
+                    ),
+                    np.round(np.clip(c + brush_size / 2 + 0.5, 0, s)).astype(
+                        int
+                    ),
                     1,
                 )
-            for i in self.dims.not_displayed:
-                slice_coord[i] = np.round(coord[i]).astype(int)
+                for c, s, brush_size in zip(coord, self.shape, brush_size_dims)
+            )
+        elif self.brush_shape == "circle":
+            slice_coord = [int(np.round(c)) for c in coord]
+            shape = self.shape
+            if not self.n_dimensional and self.ndim > 2:
+                coord = [coord[i] for i in self.dims.displayed]
+                shape = [shape[i] for i in self.dims.displayed]
+
+            sphere_dims = len(coord)
+            # Ensure circle doesn't have spurious point
+            # on edge by keeping radius as ##.5
+            radius = np.floor(self.brush_size / 2) + 0.5
+            mask_indices = self.sphere_indices(radius, sphere_dims)
+
+            mask_indices = mask_indices + np.round(np.array(coord)).astype(int)
+
+            # discard candidate coordinates that are out of bounds
+            discard_coords = np.logical_and(
+                ~np.any(mask_indices < 0, axis=1),
+                ~np.any(mask_indices > np.array(shape), axis=1),
+            )
+            mask_indices = mask_indices[discard_coords]
+
+            # Transfer valid coordinates to slice_coord,
+            # or expand coordinate if 3rd dim in 2D image
+            slice_coord_temp = [m for m in mask_indices.T]
+            if not self.n_dimensional and self.ndim > 2:
+                for j, i in enumerate(self.dims.displayed):
+                    slice_coord[i] = slice_coord_temp[j]
+                for i in self.dims.not_displayed:
+                    slice_coord[i] = slice_coord[i] * np.ones(
+                        mask_indices.shape[0], dtype=int
+                    )
```

Comment:
```suggestion
```

indices in fancy indexing are broadcast together, so this step is unnecessary — integers will get broadcast normally:

```python
In [8]: arr = np.random.random((5, 12, 12))

In [9]: indices = (np.array([5, 5, 6]), np.array([8, 9, 9]))

In [10]: arr[(1,) + indices]
Out[10]: array([0.46793242, 0.57033286, 0.64759796])

In [11]: arr[(1 * np.ones(indices[0].shape, dtype=int),) + indices]
Out[11]: array([0.46793242, 0.57033286, 0.64759796])
```

Reply:
> which fails for me,

Can you demonstrate this failure? If you look at the code I showed above, it works totally fine!

Reply:
Ah yes, that does work fine, and I see that now!

So, the problem arises however when `preserve_labels = True`, and below we have:

```python
slice_coord = tuple(sc[keep_coords] for sc in slice_coord)
```
and when one of the items in the tuple is an `int` instead of an array, it throws an exception.

that line would have to become

```python
slice_coord = tuple(sc[keep_coords] if isinstance(sc, np.ndarray) else sc for sc in slice_coord)
```
Or break it out of list comprehension really.


All of these problems melt away if we could just do 

```python
self.data[slice_coord][keep_coords] = new_label
```

downstream!!! ❗

Reply:
Ooooooooh! =) Ok, all good, let's keep the current scheme! 🎉

Reply:
Sorry 😔. 

Now that the circle brushes have tests, I welcome someone to come in and hack it up with a more elegant solution